File: src/Hedera.Hashgraph.Tests/NodeDeleteTransactionTest.cs
// SPDX-License-Identifier: Apache-2.0
using Org.Assertj.Core.Api.Assertions;
using Org.Junit.Jupiter.Api.Assertions;
using Com.Hedera.Hashgraph.Sdk.Proto;
using Io.Github.JsonSnapshot;
using Java.Time;
using Java.Util;
using Org.Junit.Jupiter.Api;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;

namespace Com.Hedera.Hashgraph.Sdk
{
    public class NodeDeleteTransactionTest
    {
        private static readonly PrivateKey TEST_PRIVATE_KEY = PrivateKey.FromString("302e020100300506032b657004220420db484b828e64b2d8f12ce3c0a0e93a0b8cce7af1bb8f39c97732394482538e10");
        private static readonly long TEST_NODE_ID = 420;
        readonly Instant TEST_VALID_START = Instant.OfEpochSecond(1554158542);
        public static void BeforeAll()
        {
            SnapshotMatcher.Start(Snapshot.AsJsonString());
        }

        public static void AfterAll()
        {
            SnapshotMatcher.ValidateSnapshots();
        }

        virtual void ShouldSerialize()
        {
            SnapshotMatcher.Expect(SpawnTestTransaction().ToString()).ToMatchSnapshot();
        }

        private NodeDeleteTransaction SpawnTestTransaction()
        {
            return new NodeDeleteTransaction().SetNodeAccountIds(Arrays.AsList(AccountId.FromString("0.0.5005"), AccountId.FromString("0.0.5006"))).SetTransactionId(TransactionId.WithValidStart(AccountId.FromString("0.0.5006"), TEST_VALID_START)).SetNodeId(TEST_NODE_ID).SetMaxTransactionFee(new Hbar(1)).Freeze().Sign(TEST_PRIVATE_KEY);
        }

        virtual void ShouldBytes()
        {
            var tx = SpawnTestTransaction();
            var tx2 = NodeDeleteTransaction.FromBytes(tx.ToBytes());
            Assert.Equal(tx2.ToString(), tx.ToString());
        }

        virtual void ShouldBytesNoSetters()
        {
            var tx = new NodeDeleteTransaction();
            var tx2 = Transaction.FromBytes(tx.ToBytes());
            Assert.Equal(tx2.ToString(), tx.ToString());
        }

        virtual void FromScheduledTransaction()
        {
            var transactionBody = SchedulableTransactionBody.NewBuilder().SetNodeDelete(NodeDeleteTransactionBody.NewBuilder().Build()).Build();
            var tx = Transaction.FromScheduledTransaction(transactionBody);
            Assert.IsType<NodeDeleteTransaction>(tx);
        }

        virtual void ConstructNodeDeleteTransactionFromTransactionBodyProtobuf()
        {
            var transactionBodyBuilder = NodeDeleteTransactionBody.NewBuilder();
            transactionBodyBuilder.SetNodeId(TEST_NODE_ID);
            var tx = TransactionBody.NewBuilder().SetNodeDelete(transactionBodyBuilder.Build()).Build();
            var nodeDeleteTransaction = new NodeDeleteTransaction(tx);
            Assert.Equal(nodeDeleteTransaction.GetNodeId(), TEST_NODE_ID);
        }

        virtual void GetSetNodeId()
        {
            var nodeDeleteTransaction = new NodeDeleteTransaction().SetNodeId(TEST_NODE_ID);
            Assert.Equal(nodeDeleteTransaction.GetNodeId(), TEST_NODE_ID);
        }

        virtual void GetSetNodeIdFrozen()
        {
            var tx = SpawnTestTransaction();
            await Assert.ThrowsAsync<InvalidOperationException>(() => tx.SetNodeId(TEST_NODE_ID));
        }

        virtual void ShouldFreezeSuccessfullyWhenNodeIdIsSet()
        {
            Instant VALID_START = Instant.OfEpochSecond(1596210382);
            AccountId ACCOUNT_ID = AccountId.FromString("0.6.9");
            var transaction = new NodeDeleteTransaction().SetNodeAccountIds(Arrays.AsList(AccountId.FromString("0.0.3"))).SetTransactionId(TransactionId.WithValidStart(ACCOUNT_ID, VALID_START)).SetNodeId(420);
            AssertThatCode(() => transaction.FreezeWith(null)).DoesNotThrowAnyException();
            Assert.Equal(transaction.GetNodeId(), 420);
        }

        virtual void ShouldThrowErrorWhenFreezingWithoutSettingNodeId()
        {
            Instant VALID_START = Instant.OfEpochSecond(1596210382);
            AccountId ACCOUNT_ID = AccountId.FromString("0.6.9");
            var transaction = new NodeDeleteTransaction().SetNodeAccountIds(Arrays.AsList(AccountId.FromString("0.0.3"))).SetTransactionId(TransactionId.WithValidStart(ACCOUNT_ID, VALID_START));
            var exception = await Assert.ThrowsAsync<InvalidOperationException>(() => transaction.FreezeWith(null));
            Assert.Equal(exception.GetMessage(), "NodeDeleteTransaction: 'nodeId' must be explicitly set before calling freeze().");
        }

        virtual void ShouldThrowErrorWhenFreezingWithZeroNodeId()
        {
            Instant VALID_START = Instant.OfEpochSecond(1596210382);
            AccountId ACCOUNT_ID = AccountId.FromString("0.6.9");
            var transaction = new NodeDeleteTransaction().SetNodeAccountIds(Arrays.AsList(AccountId.FromString("0.0.3"))).SetTransactionId(TransactionId.WithValidStart(ACCOUNT_ID, VALID_START));
            var exception = await Assert.ThrowsAsync<InvalidOperationException>(() => transaction.FreezeWith(null));
            Assert.Equal(exception.GetMessage(), "NodeDeleteTransaction: 'nodeId' must be explicitly set before calling freeze().");
        }

        virtual void ShouldFreezeSuccessfullyWithActualClientWhenNodeIdIsSet()
        {
            Instant VALID_START = Instant.OfEpochSecond(1596210382);
            AccountId ACCOUNT_ID = AccountId.FromString("0.6.9");
            var transaction = new NodeDeleteTransaction().SetNodeAccountIds(Arrays.AsList(AccountId.FromString("0.0.3"))).SetTransactionId(TransactionId.WithValidStart(ACCOUNT_ID, VALID_START)).SetNodeId(420);
            var mockClient = Client.ForTestnet();
            AssertThatCode(() => transaction.FreezeWith(mockClient)).DoesNotThrowAnyException();
            Assert.Equal(transaction.GetNodeId(), 420);
        }

        virtual void ShouldThrowErrorWhenGettingNodeIdWithoutSettingIt()
        {
            var transaction = new NodeDeleteTransaction();
            var exception = await Assert.ThrowsAsync<InvalidOperationException>(() => transaction.GetNodeId());
            Assert.Equal(exception.GetMessage(), "NodeDeleteTransaction: 'nodeId' has not been set");
        }

        virtual void ShouldThrowErrorWhenSettingNegativeNodeId()
        {
            var transaction = new NodeDeleteTransaction();
            var exception = await Assert.ThrowsAsync<ArgumentException>(() => transaction.SetNodeId(-1));
            Assert.Equal(exception.GetMessage(), "NodeDeleteTransaction: 'nodeId' must be non-negative");
        }

        virtual void ShouldAllowSettingNodeIdToZero()
        {
            var transaction = new NodeDeleteTransaction().SetNodeId(0);
            Assert.Equal(transaction.GetNodeId(), 0);
        }
    }
}

File: src/test/java/com/hedera/hashgraph/sdk/NodeDeleteTransactionTest.java
// SPDX-License-Identifier: Apache-2.0
package com.hedera.hashgraph.sdk;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatCode;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.hedera.hashgraph.sdk.proto.NodeDeleteTransactionBody;
import com.hedera.hashgraph.sdk.proto.SchedulableTransactionBody;
import com.hedera.hashgraph.sdk.proto.TransactionBody;
import io.github.jsonSnapshot.SnapshotMatcher;
import java.time.Instant;
import java.util.Arrays;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

public class NodeDeleteTransactionTest {

    private static final PrivateKey TEST_PRIVATE_KEY = PrivateKey.fromString(
            "302e020100300506032b657004220420db484b828e64b2d8f12ce3c0a0e93a0b8cce7af1bb8f39c97732394482538e10");

    private static final long TEST_NODE_ID = 420;

    final Instant TEST_VALID_START = Instant.ofEpochSecond(1554158542);

    @BeforeAll
    public static void beforeAll() {
        SnapshotMatcher.start(Snapshot::asJsonString);
    }

    @AfterAll
    public static void afterAll() {
        SnapshotMatcher.validateSnapshots();
    }

    @Test
    void shouldSerialize() {
        SnapshotMatcher.expect(spawnTestTransaction().toString()).toMatchSnapshot();
    }

    private NodeDeleteTransaction spawnTestTransaction() {
        return new NodeDeleteTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), TEST_VALID_START))
                .setNodeId(TEST_NODE_ID)
                .setMaxTransactionFee(new Hbar(1))
                .freeze()
                .sign(TEST_PRIVATE_KEY);
    }

    @Test
    void shouldBytes() throws Exception {
        var tx = spawnTestTransaction();
        var tx2 = NodeDeleteTransaction.fromBytes(tx.toBytes());
        assertThat(tx2.toString()).isEqualTo(tx.toString());
    }

    @Test
    void shouldBytesNoSetters() throws Exception {
        var tx = new NodeDeleteTransaction();
        var tx2 = Transaction.fromBytes(tx.toBytes());
        assertThat(tx2.toString()).isEqualTo(tx.toString());
    }

    @Test
    void fromScheduledTransaction() {
        var transactionBody = SchedulableTransactionBody.newBuilder()
                .setNodeDelete(NodeDeleteTransactionBody.newBuilder().build())
                .build();

        var tx = Transaction.fromScheduledTransaction(transactionBody);

        assertThat(tx).isInstanceOf(NodeDeleteTransaction.class);
    }

    @Test
    void constructNodeDeleteTransactionFromTransactionBodyProtobuf() {
        var transactionBodyBuilder = NodeDeleteTransactionBody.newBuilder();

        transactionBodyBuilder.setNodeId(TEST_NODE_ID);

        var tx = TransactionBody.newBuilder()
                .setNodeDelete(transactionBodyBuilder.build())
                .build();
        var nodeDeleteTransaction = new NodeDeleteTransaction(tx);

        assertThat(nodeDeleteTransaction.getNodeId()).isEqualTo(TEST_NODE_ID);
    }

    @Test
    void getSetNodeId() {
        var nodeDeleteTransaction = new NodeDeleteTransaction().setNodeId(TEST_NODE_ID);
        assertThat(nodeDeleteTransaction.getNodeId()).isEqualTo(TEST_NODE_ID);
    }

    @Test
    void getSetNodeIdFrozen() {
        var tx = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> tx.setNodeId(TEST_NODE_ID));
    }

    @Test
    @DisplayName("should freeze successfully when nodeId is set")
    void shouldFreezeSuccessfullyWhenNodeIdIsSet() {
        final Instant VALID_START = Instant.ofEpochSecond(1596210382);
        final AccountId ACCOUNT_ID = AccountId.fromString("0.6.9");

        var transaction = new NodeDeleteTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.3")))
                .setTransactionId(TransactionId.withValidStart(ACCOUNT_ID, VALID_START))
                .setNodeId(420);

        assertThatCode(() -> transaction.freezeWith(null)).doesNotThrowAnyException();
        assertThat(transaction.getNodeId()).isEqualTo(420);
    }

    @Test
    @DisplayName("should throw error when freezing without setting nodeId")
    void shouldThrowErrorWhenFreezingWithoutSettingNodeId() {
        final Instant VALID_START = Instant.ofEpochSecond(1596210382);
        final AccountId ACCOUNT_ID = AccountId.fromString("0.6.9");

        var transaction = new NodeDeleteTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.3")))
                .setTransactionId(TransactionId.withValidStart(ACCOUNT_ID, VALID_START));

        var exception = assertThrows(IllegalStateException.class, () -> transaction.freezeWith(null));
        assertThat(exception.getMessage())
                .isEqualTo("NodeDeleteTransaction: 'nodeId' must be explicitly set before calling freeze().");
    }

    @Test
    @DisplayName("should throw error when freezing with nodeId null")
    void shouldThrowErrorWhenFreezingWithZeroNodeId() {
        final Instant VALID_START = Instant.ofEpochSecond(1596210382);
        final AccountId ACCOUNT_ID = AccountId.fromString("0.6.9");

        var transaction = new NodeDeleteTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.3")))
                .setTransactionId(TransactionId.withValidStart(ACCOUNT_ID, VALID_START));

        var exception = assertThrows(IllegalStateException.class, () -> transaction.freezeWith(null));
        assertThat(exception.getMessage())
                .isEqualTo("NodeDeleteTransaction: 'nodeId' must be explicitly set before calling freeze().");
    }

    @Test
    @DisplayName("should freeze successfully with actual client when nodeId is set")
    void shouldFreezeSuccessfullyWithActualClientWhenNodeIdIsSet() {
        final Instant VALID_START = Instant.ofEpochSecond(1596210382);
        final AccountId ACCOUNT_ID = AccountId.fromString("0.6.9");

        var transaction = new NodeDeleteTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.3")))
                .setTransactionId(TransactionId.withValidStart(ACCOUNT_ID, VALID_START))
                .setNodeId(420);

        var mockClient = Client.forTestnet();

        assertThatCode(() -> transaction.freezeWith(mockClient)).doesNotThrowAnyException();
        assertThat(transaction.getNodeId()).isEqualTo(420);
    }

    @Test
    @DisplayName("should throw error when getting nodeId without setting it")
    void shouldThrowErrorWhenGettingNodeIdWithoutSettingIt() {
        var transaction = new NodeDeleteTransaction();

        var exception = assertThrows(IllegalStateException.class, () -> transaction.getNodeId());
        assertThat(exception.getMessage()).isEqualTo("NodeDeleteTransaction: 'nodeId' has not been set");
    }

    @Test
    @DisplayName("should throw error when setting negative nodeId")
    void shouldThrowErrorWhenSettingNegativeNodeId() {
        var transaction = new NodeDeleteTransaction();

        var exception = assertThrows(IllegalArgumentException.class, () -> transaction.setNodeId(-1));
        assertThat(exception.getMessage()).isEqualTo("NodeDeleteTransaction: 'nodeId' must be non-negative");
    }

    @Test
    @DisplayName("should allow setting nodeId to zero")
    void shouldAllowSettingNodeIdToZero() {
        var transaction = new NodeDeleteTransaction().setNodeId(0);
        assertThat(transaction.getNodeId()).isEqualTo(0);
    }
}
